### PR TITLE
2686: Proposal Process Design Doc

### LINKED ISSUE #2686: Introduce proposal process to Tanzu Community Edition
## Abstract

The Tanzu Community Edition project needs a proposal process that can foster:

* Visibility into proposed changes to the project
* Feedback from maintainers, community members, and end-users
* Reaching decisions on proposed changes in a timely manner

## Proposal

Detailed proposal available in a design doc:

* [PR](https://github.com/vmware-tanzu/community-edition/pull/2687)
* [Rendered View](https://github.com/vmware-tanzu/community-edition/blob/proposal-process/docs/designs/2686-proposal-process.md)

## Alternatives Considered

None.

### PR DESCRIPTION
**Pull request by joshrosso**
This design doc covers the proposal process for Tanzu Community Edition.

It attempts to foster:

* Visibility into proposed changes to the project
* Feedback from maintainers, community members, and end-users
* Reaching decisions on proposed changes in a timely manner

Relates to: https://github.com/vmware-tanzu/community-edition/issues/2686

### CODE REVIEW

File: docs/designs/2686-proposal-process.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+guidelines for how to create a proposal.
+
+## The Proposal Process
+
+The proposal process is the process for reviewing a proposal and reaching a decision about whether
+to accept or decline the proposal.
+
+1. The proposal author [creates a brief issue](https://golang.org/issue/new) describing the proposal.
+
+   > A design doc is **not** required at this point.
```
Comment (stmcginnis) [inline docs/designs/2686-proposal-process.md:18]: Maybe soften this with saying a design doc is welcome if they want to provide one, but not required at this point.

File: docs/designs/2686-proposal-process.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+1. A discussion on the issue tracker aims to triage the proposal into one of three outcomes:
+
+    * Accepted (`proposal/accepted` label)
+    * Declined (`proposal/declined` label)
+    * Needs Design Doc (`proposal/needs-design-doc` label)
+
+   > If the proposal is Accepted or Declined, the process is done.
+
+1. Maintainers determine whether a design doc is required. If it is, a `proposal/needs-design-doc`
+   label.
```
Comment (stmcginnis) [inline docs/designs/2686-proposal-process.md:29]: s/label./label will be applied./

Comment (joshrosso) [inline docs/designs/2686-proposal-process.md:29]:
```suggestion
   label is added.
```

File: docs/designs/2686-proposal-process.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+1. Maintainers review open proposals weekly to move them forward.
+
+## Design Documents
+
+When a design doc is requested, a proposal author should:
+
+1. Create a **branch** on their **fork** of the `community-edition` repository.
+
+1. Create the file `docs/designs/${GH_ISSUE_NUMBER}-${SHORT_PROPOSAL_TITLE}.md`.
```
Comment (jpmcb) [inline docs/designs/2686-proposal-process.md:42]: Will _all_ proposals, past and present, be stored here? Or will be at some latter time introduce a way for these to get trimmed? I'm ok with a long directory of proposals, but might get unruly.

File: docs/designs/template.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+< Provide a short summary of what would be achieved and why is is of value >
+
+< If user stories are relevant, add them here >
+
+## Proposal
+
+< Provide a detailed proposal of **what** would be done and **how**; feel free to add multiple
+subsections, diagrams, and config/code snippets >
+
```
Comment (stmcginnis) [inline docs/designs/template.md:18]: Might be nice to have a section that explicitly calls out use cases or scenarios that justify why this change is needed. I've seen that work well to frame the discussion and help illustrate why certain choices are being made.

File: docs/designs/2686-proposal-process.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+Tanzu Community Edition relies on design proposals to assess the viability of non-trivial changes
+to the project. This document covers details on the proposal process, which can also be used as
+guidelines for how to create a proposal.
+
+## The Proposal Process
+
+The proposal process is the process for reviewing a proposal and reaching a decision about whether
+to accept or decline the proposal.
+
+1. The proposal author [creates a brief issue](https://golang.org/issue/new) describing the proposal.
```
Comment (stmcginnis) [inline docs/designs/2686-proposal-process.md:16]: Should we get an issue template added specifically for this?

File: docs/designs/2686-proposal-process.md (hunk 3 above)
Comment (joshrosso) [reply docs/designs/2686-proposal-process.md:42]: Unfortunately this will grow over time. As talked about in our review session, there _may_ be a future state where we move proposals out into it's own repository. But for the time being, we'll incubate this process in our own.

File: docs/designs/2686-proposal-process.md (hunk 1 above)
Comment (joshrosso) [reply docs/designs/2686-proposal-process.md:18]: Agreed.

Also: we should encourage that POCs and code **are** ok! You don't need to wait. But you should know that you're running the risk of the proposal getting declined and the code going to waste.

File: docs/designs/template.md (hunk 4 above)
Comment (joshrosso) [reply docs/designs/template.md:18]: Sub-level of abstract around user-stories and why the change should be made.

File: docs/designs/template.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Proposal Title
+
+* Proposal: < GITHUB ISSUE URL >
+* See Also:
+  * < RELEVANT URL >
+  * < RELEVANT URL >
+
+## Abstract
+
+< Provide a short summary of what would be achieved and why is is of value >
```
Comment (stmcginnis) [inline docs/designs/template.md:10]: s/is is/it is/

**Comment by joshrosso** [comment]
We need a decision around where PM fits into this **technical** proposal process.

File: docs/designs/2686-proposal-process.md (hunk 5 above)
Comment (joshrosso) [reply docs/designs/2686-proposal-process.md:16]: As talked about in review today, we're going to try for a while without a specific issue template and see if converting existing issues (e.g. `kind/feature`) to a proposal is adequate.

File: docs/designs/template.md (hunk 4 above)
Comment (joshrosso) [reply docs/designs/template.md:21]: Add a `Why` subsection to the abstract.

If this satisfies the request, please resolve this comment. Thanks!

File: docs/designs/2686-proposal-process.md (hunk 1 above)
Comment (joshrosso) [reply docs/designs/2686-proposal-process.md]: Added verbiage to soften this and also talk a bit about starting an implementation before a proposal is accepted.

**Review by davidvonthenen — DISMISSED** [review]
this looks good to me

File: docs/designs/2686-proposal-process.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+   > A design doc is **not** required at this point; although proposal authors
+   > are welcome to create one.
+
```
Comment (jpmcb) [inline docs/designs/2686-proposal-process.md:20]: Apparently, MD lint doesn't like a blank line here inside the block quote:
```
docs/designs/2686-proposal-process.md:20 MD028/no-blanks-blockquote Blank line inside blockquote
```

Question: do we want to enable MD linter inside of the `docs/designs/` directory or should we disable it, giving the author the freedom to format it how they will?

Comment (stmcginnis) [reply docs/designs/2686-proposal-process.md:20]: I do think it's useful having the mdlinting to make sure the markdown is well formatted and will render properly.

File: docs/designs/template.md (hunk 4 above)
Comment (dims) [reply docs/designs/template.md:21]: @stmcginnis @joshrosso i'll taking the liberty of adding more things in the template.

**Comment by qnetter** [comment]
When a proposal is being reviewed and we are in the RFC period, we should set a deadline for input, so nobody waits and misses the window.

**Comment by dims** [comment]
@joshrosso i can't seem to suggest changes, i've augmented the template markdown with a bunch of things we learned from k/k and Cluser-API proposal process. For example a machine parsable markdown at the top of the file. Explicit sections for user stories, a concrete table of contents etc.

I've made sure i copied all the current contents as well (for example i did not have a Compatibility section before). Also there are step-by-step instructions and tid-bits for authors.

One point of contention would be ... is this too much? (example, i am proposing a larger set of labels for status than just the 3 we had in this PR) 

Things i looked at were:
- https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/YYYYMMDD-template.md
- https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template

https://gist.github.com/dims/b9cd0c191c785292a8efad34030faf48

Happy to iterate.

Weird that i can't suggest changes! let me try tracking this down.

**Comment by dims** [comment]
Looks like i need to be in the list of "Reviewers" for me to suggest changes ...

File: docs/designs/2686-proposal-process.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
+# Proposing Changes to Tanzu Community Edition
+
+* Proposal: [https://github.com/vmware-tanzu/community-edition/issues/2686](https://github.com/vmware-tanzu/community-edition/issues/2686)
```
Comment (dims) [inline docs/designs/2686-proposal-process.md:3]: testing something

**Comment by dims** [comment]
so looks like i can suggest lines (or a para) and can't edit a full file. So we should probably incorporate bits from the gist into the template.md in this PR. thanks @joshrosso

**Comment by joshrosso** [comment]
Final TODOs: 

* Incorp most of caep template in to our template (marking some as optional and removing TOC requirement
* Add to review process having upstream/dependencies involved in approval when relevant
* Add to review process that when reviews begin, we declare an RFC window, and we put that in the proposal issue (maybe slack reminders, etc)

**Comment by joshrosso** [comment]
> When a proposal is being reviewed and we are in the RFC period, we should set a deadline for input, so nobody waits and misses the window.

I've addressed this by including the following verbiage:

> The [TCE
> maintainers](https://github.com/vmware-tanzu/community-edition/blob/main/MAINTAINERS.md#maintainers)
> meet weekly to review proposals that have not been Approved or Declined. A
> summary of the discussion for each proposal will be added to the proposal issue
> in GitHub. The outcome of these reviews could be:
>
> * No changes
> * Requests for more details
> * Approving a proposal
> * Declining a proposal

**Comment by joshrosso** [comment]
I have more work tomorrow to clean-up the template and fix MD lint issues in it.

File: docs/designs/2686-proposal-process.md (hunk 7 above)
Comment (joshrosso) [reply docs/designs/2686-proposal-process.md:21]: I agree.

I added the following to the design doc section:

```
_note: PRs for design docs are linted using
[markdownlint](https://github.com/markdownlint/markdownlint), you can run it
locally or use `make check`._
```

**Review by davidvonthenen — APPROVED** [review]

**Review by seemiller — APPROVED** [review]
LGTM

**Review by stmcginnis — APPROVED** [review]
